File: find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target) {
        if(nums==null || nums.Length ==0) return new int[]{-1,-1};
       int[] targetrange = new int[] {-1,-1}; 
        for(int i=0;i<nums.Length;i++)
        {
            if(nums[i]==target)
            {
                targetrange[0]=i;
                break;
            }
        }
        
        if(targetrange[0]==-1)
        {
            return targetrange;
        }
        
        for(int j=nums.Length-1;j>=0;j--)
        {
            if(nums[j]==target)
            {
                targetrange[1]=j;
                break;
            }
        }
        return targetrange;
        
    }
}

File: find-first-and-last-position-of-element-in-sorted-array.cpp
/*
The code finds the first and last occurrences of a target element in a sorted array using two binary searches.
The findBound function is used for both searches, with a parameter to determine whether to find the first or last occurrence.
The search space is adjusted accordingly in each binary search until the target element is found.
Time Complexity: O(log n)
Space Complexity: O(1)
*/
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int firstOccurrence = findBound(nums, target, 1);
        vector<int> p;
        if (firstOccurrence == -1) {
            p.push_back(-1);
            p.push_back(-1);
            return p;
        }
        
        int lastOccurrence = findBound(nums, target, 0);
        vector<int> k;
        k.push_back(firstOccurrence);
        k.push_back(lastOccurrence);
        return k;
    }

    int findBound(vector<int>& nums, int target, int isFirst) {
        int N = nums.size();
        int begin = 0, end = N - 1;
        
        while (begin <= end) {
            
            int mid = (begin + end) / 2;
            
            if (nums[mid] == target) {
                
                if (isFirst) {
                    
                    if (mid == begin || nums[mid - 1] != target) {
                        return mid;
                    }
                    
                    end = mid - 1;
                    
                } else {
                    
                    if (mid == end || nums[mid + 1] != target) {
                        return mid;
                    }
                    
                    begin = mid + 1;
                }
                
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                begin = mid + 1;
            }
        }
        
        return -1;
    }
};